refactor: Guard TxRequestTracker by its own lock instead of cs_main

dergoegge opened the pull request:
I don't see the need to have the `TxRequestTracker` guarded by `cs_main` which would also be more in line with our developer docs.

From `developer-notes.md`:

```
Re-architecting the core code so there are better-defined interfaces between
the various components is a goal, with any necessary locking done by the
components (e.g. see the self-contained FillableSigningProvider class and its
cs_KeyStore lock for example).
```

This PR gives `TxRequestTracker` its own mutex, thereby removing the need to guard `PeerManagerImpl::m_txrequest` using `cs_main`.

dergoegge committed:
[net processing] Guard TxRequestTracker internally using its own mutex

dergoegge committed:
[net processing] Reduce cs_main scope for m_txrequest

m_txrequest is now guarded by TxRequestTracker's internal mutex.

jnewbery commented:
Concept ACK. Originally suggested in https://github.com/bitcoin/bitcoin/pull/19988#discussion_r497882360

fanquake commented on src/txrequest.cpp:
```diff
@@ -534,8 +538,10 @@ class TxRequestTracker::Impl {
-    void DisconnectedPeer(NodeId peer)
+    void DisconnectedPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_txrequest_mutex)
     {
+		LOCK(m_txrequest_mutex);
```
linter is sad about tabs being used here

fanquake reviewed (commented):
concept ack - restarted the *SAN job as that was an unrelated apt failure.

ajtowns reviewed (commented):
If there's to be a separate lock, I'm not convinced it makes sense to push the lock all the way into the `Impl` class -- having `m_impl GUARDED_BY(m_impl_mutex)` in `TxRequestTracker`, or having `m_txrequest GUARDED_BY(m_txrequest_mutex)` seem more sensible to me, and the latter allows you to take the lock outside of the various for loops, rather than repeatedly taking it and dropping it on each iteration.

I think the arguments from the original PR still apply though -- this doesn't meaningfully allow any more parallelism / reduce blocking as far as I can see.

dergoegge commented:
Thanks @jnewbery for linking the previous discussion! I was looking through the old PRs yesterday and didn't come across that comment.

> If there's to be a separate lock, I'm not convinced it makes sense to push the lock all the way into the Impl class -- having m_impl GUARDED_BY(m_impl_mutex) in TxRequestTracker, or having m_txrequest GUARDED_BY(m_txrequest_mutex) seem more sensible to me, and the latter allows you to take the lock outside of the various for loops, rather than repeatedly taking it and dropping it on each iteration.

@ajtowns I'm not sure if I am on the same page about this but I am willing to be convinced. I used `AddrMan` (and the dev notes) as a reference for choosing the current design (i.e. putting the lock inside the `Impl` class).
* Would you be in favor of refactoring our existing modules to externalize their locks? (e.g. `AddrManImpl::cs`)
* Should our developer docs be updated? (I think our lock conventions are a bit all over the place, similar to general coding style, so maybe it helps to clarify our preferences?)

> I think the arguments from the original PR still apply though -- this doesn't meaningfully allow any more parallelism / reduce blocking as far as I can see.

I have been looking at reducing `cs_main` usage within net processing because I don't think that our networking code should be locking `cs_main` (our main validation lock) as much as it currently does. Reducing the scope of `cs_main` to validation seems like a good direction to be heading in (especially w.r.t. the kernel project). This PR seemed like an easy small step in that direction (similar to #26140) as it does remove two `LOCK(cs_main)` call sites.

dergoegge commented on src/net_processing.cpp:
```diff
@@ -1399,6 +1399,7 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, const Peer& peer)
 void PeerManagerImpl::AddTxAnnouncement(const CNode& node, const GenTxid& gtxid, std::chrono::microseconds current_time)
 {
     AssertLockHeld(::cs_main); // For m_txrequest
+
```
Removed the comment

dergoegge commented on src/txrequest.cpp:
```diff
@@ -307,6 +307,8 @@ GenTxid ToGenTxid(const Announcement& ann)
 
 /** Actual implementation for TxRequestTracker's data structure. */
 class TxRequestTracker::Impl {
+    mutable Mutex m_txrequest_mutex;
```
Done.

ajtowns commented:
> @ajtowns I'm not sure if I am on the same page about this but I am willing to be convinced. I used `AddrMan` (and the dev notes) as a reference for choosing the current design (i.e. putting the lock inside the `Impl` class).
> 
>     * Would you be in favor of refactoring our existing modules to externalize their locks? (e.g. `AddrManImpl::cs`)

In general, I think it's better to leave things alone if we're not making things substantially more reliable/simple/efficient.

For locking, I think the best approach on all of those axes is "having a reference to an object means you can manipulate it; if you can't manipulate it, you don't have a reference to in the first place"; but that's often hard to achieve.

Maybe compare with what @vasild's proposing in #25390 -- with that you'd just say `Synced<TxRequestTracker> m_txrequest` and write either `m_txrequest->Foo();` to do a single operation that takes then releases the lock, or `{ auto proxy = *m_txrequest; proxy->Foo(); proxy->Bar(); }` to do multiple operations while the lock's held. Meanwhile, if you're not doing threading (like in the unit/fuzz tests) you just allocate a `TxRequestTracker` directly, and don't worry about locks at all. (Unfortunately, I don't think the implementation there quite works right/clang isn't clever enough to properly understand it, and I haven't been able to come up with a better one. Err, except, maybe...)

> I have been looking at reducing `cs_main` usage within net processing because I don't think that our networking code should be locking `cs_main` (our main validation lock) as much as it currently does. Reducing the scope of `cs_main` to validation seems like a good direction to be heading in (especially w.r.t. the kernel project). This PR seemed like an easy small step in that direction (similar to #26140) as it does remove two `LOCK(cs_main)` call sites.

Yeah... Kind of feel like it's probably better spending coding/review time on the hard steps though? For `cs_main`, having dedicated, non-global, locks for blockstorage, block indexes, and coin states, so that you don't need `cs_main` there, seems like the priority. For net, getting rid of `CNodeState` entirely and passing around `Peer&` objects seems worthwhile, and I think maybe adding an opaque `PeerRef` to `CNode` would allow avoiding the extra `map` and ensure the `Peer` object doesn't get deallocated before the `CNode` object does... We could also perhaps have more things under "only accessed by the message processing thread" by having that thread make read-only copies of the data available to other threads in advance, which would then reduce contention...

hebasto commented:
Concept ACK.

hebasto commented:
> Maybe compare with what @vasild's proposing in #25390 -- with that you'd just say `Synced<TxRequestTracker> m_txrequest` and write either `m_txrequest->Foo();` to do a single operation that takes then releases the lock, or `{ auto proxy = *m_txrequest; proxy->Foo(); proxy->Bar(); }` to do multiple operations while the lock's held.

It makes me think that such an issue should be resolved at `TxRequestTracker` class's API level. An object which self maintains its state in multi-threading environment is preferable (less error prone, easy to reason about etc).

dergoegge committed:
[txrequest] Add method for removing multiple transaction requests at once

vasild reviewed (approved):
ACK 2ac77c4b5ef095846924be3c91e02967e8177abf

This PR adds an internal lock and acquires it inside the methods that touch the relevant variables. This would allow finer grained control - e.g. to lock the mutex only for some part of a method, to improve concurrency. In this PR, however, we don't do that - we lock the mutex for the entire duration of the methods, which is the same as having an external mutex outside, like @ajtowns mentioned above:

> ... or having `m_txrequest GUARDED_BY(m_txrequest_mutex)` ... allows you to take the lock outside of the various for loops, rather than repeatedly taking it and dropping it on each iteration.

Which way is better would depend on how this is used by multiple threads. Because it is not, both approaches look equally good (or equally bad) _now_.

Thanks, @ajtowns, for mentioning https://github.com/bitcoin/bitcoin/pull/25390. This PR can be achieved by just:

```diff
-    TxRequestTracker m_txrequest;
+    Synced<TxRequestTracker> m_txrequest;
```
and mechanically replacing `.` with `->` (we don't need to hold the lock across multiple method calls). That is virtually one line of change. See this comment: https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1169788319 looks like this PR is doing the "Lots of repetitions" case.

Riahiamirreza commented on src/txrequest.cpp:
```diff
@@ -306,23 +306,28 @@ GenTxid ToGenTxid(const Announcement& ann)
 
 /** Actual implementation for TxRequestTracker's data structure. */
-class TxRequestTracker::Impl {
+class TxRequestTracker::Impl
```
I know this question is not in the scope of this PR but appreciate the answer. Why there is a separation between `TxRequestTracker` and it's implementation? The comments says: ` Avoid littering this header file with implementation details.`
Is the clarity and readability of code, the whole point of this separation? If yes, is this a common design pattern?

ajtowns replied:
That's exactly the point. It's called the "pimpl pattern" (pointer to impl class) -- https://cpppatterns.com/patterns/pimpl.html

dergoegge replied:
We also had a review club on this a while ago, if you are interested: https://bitcoincore.reviews/22950

vasild replied:
Yeah, also here: https://en.cppreference.com/w/cpp/language/pimpl "This technique is used to construct C++ library interfaces with stable ABI and to reduce compile-time dependencies."

What it does for us in practice is to reduce _re_-compilation times - if everything is compiled and some private member is changed then only the `cpp` file will be recompiled instead of all files that include the `h` file.

Improving clarity and readability is not mentioned anywhere in any of the pimpl articles I checked.

MarcoFalke commented:
Could also move it out of the cs_main scope in `FinalizeNode`?

dergoegge commented on src/txrequest.cpp:
```diff
@@ -306,23 +306,28 @@ GenTxid ToGenTxid(const Announcement& ann)
+class TxRequestTracker::Impl
+{
+private:
+    mutable Mutex m_txrequest_mutex;
```
Renamed.

I have seen nits on other PRs suggesting the exact reverse, which is why i named it this way in the first place, hoping to avoid the nit 😅 But i actually kind of agree that its redundant, so i shouldn't have preemptively tried to avoid the nit.

dergoegge commented:
Rebased, added a method to remove multiple transactions from the tracker at once (to avoid locking the internal lock over and over again), and renamed `m_txrequest_mutex` -> `m_mutex`.

vasild commented on src/net_processing.cpp:
```diff
@@ -1819,13 +1820,8 @@ void PeerManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock
-        }
-    }
+
+    m_txrequest.ForgetTxs(Span{pblock->vtx});
```
nit:

```suggestion
    m_txrequest.ForgetTxs(pblock->vtx);
```

vasild reviewed (commented):
~~ACK 1d72d4206df579483f8297acd163611fc294378e~~ See [below](https://github.com/bitcoin/bitcoin/pull/26151#issuecomment-1286897751).

vasild commented:
> Could also move it out of the cs_main scope in `FinalizeNode`?

I am not sure about it. In `PeerManagerImpl::FinalizeNode()`:

```cpp
LOCK(cs_main);
...
m_txrequest.DisconnectedPeer(nodeid);
...
assert(m_txrequest.Size() == 0);
```

hmm, wait! is that a bug in this PR? If somebody modifies (adds to) `m_txrequest` after `DisconnectedPeer()` and before the `assert()` then the `assert()` will be triggered. And the point of this PR is to be able to access (add to) `m_txrequest` without `cs_main` from anywhere :bomb: :fire:

Maybe return the size from `DisconnectedPeer()` after the deletion and later assert that the returned size was `0` in `FinalizeNode()`?

dergoegge commented:
> hmm, wait! is that a bug in this PR? If somebody modifies (adds to) m_txrequest after DisconnectedPeer() and before the assert() then the assert() will be triggered. And the point of this PR is to be able to access (add to) m_txrequest without cs_main from anywhere

I don't think it is currently but it almost is! The `assert(m_txrequest.Size() == 0);` is gated by `if (m_node_state.empty())` so no other peer can modify (add txs) `m_txrequest`.

I will still change this as that seems a little brittle wrt future changes.

vasild commented:
How? It seems that the assumption in `FinalizeNode()` is that `m_node_state` is modified together/atomically with `m_txrequest`.

Yes, the current code is fine, but then it is fine even without this PR. The aim is to make it future proof.

Returning the size from `DisconnectedPeer()` like I suggested above seems to have its own (theoretical) flaw - it could return `1`, afterwards `m_node_state` could become empty by another thread, we would enter the `if` and the assert would fail because 1 != 0.

Maybe just delete the assert? Or expose the mutex of `m_txrequest`, lock it before `DisconnectedPeer()` and unlock it after the assert :question:

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1395,7 +1395,8 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, const Peer& peer)
 void PeerManagerImpl::AddTxAnnouncement(const CNode& node, const GenTxid& gtxid, std::chrono::microseconds current_time)
 {
-    AssertLockHeld(::cs_main); // For m_txrequest
+    AssertLockHeld(::cs_main);
```
Not sure if this is "correct". While `m_txrequest` has an internal mutex to guard against UB, the internal mutex does nothing to guard the processing logic. A mutex different from the internal one is still needed here to guard against several threads calling into this function at the same time.

dergoegge replied:
Hm I think you're right. Also seems similar to what Vasil [mentioned](https://github.com/bitcoin/bitcoin/pull/26151#issuecomment-1286897751), when we call methods on `m_txrequest` but expect the internal state between those calls not to change, then the internal mutex won't help us. afaict the changes here don't break anything because `cs_main` is still held in these places but I am trying to seperate `m_txrequest` from `cs_main` so `cs_main` should not be needed for any of this after the PR. Will mark as draft until I figure out how to address this. (Seems like the only way to address this is to change the `TxRequestTracker` interface)

MarcoFalke replied:
Yeah, an alternative would be to replace cs_main with g_msgproc_mutex?

vasild replied:
Or expose the mutex outside of `TxRequestTracker` and lock for longer duration in the net processing code. I.e. have `TxRequestTracker m_txrequest GUARDED_BY(m_its_own_mutex);` as suggested by @ajtowns in https://github.com/bitcoin/bitcoin/pull/26151#pullrequestreview-1116661944

dergoegge commented:
Closing this for now, can be marked up for grabs.

https://github.com/bitcoin/bitcoin/pull/26151#discussion_r1003216463 needs to be addressed for this to move forward. Imo, the interface of `TxRequestTracker` should change to internally enforce the `MAX_PEER_TX_ANNOUNCEMENTS` and `MAX_PEER_TX_REQUEST_IN_FLIGHT` limits.
